pkg/controller/controller.go

```diff
@@ -561,6 +563,31 @@ func (p *csiProvisioner) ProvisionExt(options controller.ProvisionOptions) (*v1.
 		return nil, controller.ProvisioningInBackground, capErr
 	}
 
+	if options.PVC.Spec.DataSource != nil && (rc.clone || rc.snapshot) {
+		contentSource := rep.GetVolume().ContentSource
+		if contentSource == nil {
+			sourceErr := fmt.Errorf("volume content source missing")
+			delReq := &csi.DeleteVolumeRequest{
+				VolumeId: rep.GetVolume().GetVolumeId(),
+			}
+			delReq.Secrets = provisionerCredentials
+			ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+			defer cancel()
+			for i := 0; i < deleteVolumeRetryCount; i++ {
+				_, err = p.csiClient.DeleteVolume(ctx, delReq)
+				if err == nil {
+					break
+				}
+				klog.V(3).Infof("%v. cleanup of volume %s failed", sourceErr, pvName)
+			}
+			if err != nil {
+				sourceErr = fmt.Errorf("%v. cleanup of volume %s failed, volume is orphaned: %v", sourceErr, pvName, err)
+			}
+			klog.Error(sourceErr)
```

Comment:
I don't think we need to log here, it should be logged by the caller

Reply:
@msau42 Thanks for your comments